src/estimators/same_key_aggregator.py

```diff
@@ -94,54 +98,126 @@ def assert_compatible(self, other):
         there's no max_freq.
-      noiser_class: a class of noiser indicating the distribution of noise.
-      epsilon: total privacy budget for a run of frequency estimation. No noise
-        is added when epsilon == np.Inf.
-      epsilon_split: The proportion of total privacy budget that is assigned to
-        the estimate of 1+ reach. The remaining privacy budget is assigned to
-        the frequency histogram among effective keys.
+      reach_noiser_class: a class of noise to be added to one_plus_reach. See
+        wfa_cardinality_estimation_evaluation_framework.estimators.estimator_noisers
+        for candidate classes.
+      frequency_noiser_class: a class of noise to be added to active_key_histogram.
+        See wfa_cardinality_estimation_evaluation_framework.estimators.estimator_noisers
+        for candidate classes.
+      reach_epsilon: DP-epsilon for the noise on one_plus_reach.
+      frequency_epsilon: DP-epsilon for the noise on active_key_histogram.
+      reach_delta: DP-delta for the noise on one_plus_reach. Set as 0 if
+        reach_noise_class guarantees purely epsilon-DP.
+      frequency_delta: DP-delta for the noise on active_key_histogram. Set as 0
+        if frequency_noise_class guarantees purely epsilon-DP.
+      reach_noiser_kwargs: a dictionary of other kwargs to be specified in
+        reach_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
+      frequency_noiser_kwargs: a dictionary of other kwargs to be specified in
+        frequency_noiser_class. For example, it can be an empty dictionary, or
+        {'random_state': <a np.random.RandomState>}.
     """
     self.max_freq = max_freq
-    self.one_plus_reach_noiser = None
-    self.histogram_noiser = None
-    if noiser_class is not None:
-      assert epsilon_split > 0 and epsilon_split < 1, (
-          'In StandardizedHistogramEstimator, epsilon_split must be >0 and <1.')
-      # self.one_plus_reach_noiser is the noiser on the estimate of 1+ reach.
-      self.one_plus_reach_noiser = noiser_class(epsilon=epsilon * epsilon_split)
-      # self.histogram_noiser is the noiser on the frequency histogram among
-      # effective keys
-      # For simplicity, suppose the two noisers share the same noiser_class.
-      self.histogram_noiser = noiser_class(
-          epsilon=epsilon * (1 - epsilon_split))
+    self.reach_epsilon = (np.Inf if reach_noiser_class is None
+                          else reach_epsilon)
+    self.frequency_epsilon = (np.Inf if frequency_noiser_class is None
+                              else frequency_epsilon)
+    self.reach_delta = reach_delta
+    self.frequency_delta = frequency_delta
+    self.one_plus_reach_noiser, ignore_delta = (
+        StandardizedHistogramEstimator.define_noiser(
+            reach_noiser_class,
+            reach_epsilon, reach_delta, reach_noiser_kwargs))
+    if ignore_delta:
```

Comment:
A complexity here is that for frequency noiser, it will return per_bucket_delta and we should multiply it by 2 to save the whole histogram delta. While for reach we directly save the returned delta.

Reply:
ok